services/venice-controller/src/main/java/com/linkedin/venice/controller/kafka/protocol/serializer/NewSemanticUsageValidator.java

```diff
@@ -0,0 +1,253 @@
+package com.linkedin.venice.controller.kafka.protocol.serializer;
+
+import com.linkedin.avroutil1.compatibility.AvroCompatibilityHelper;
+import com.linkedin.venice.utils.AvroSchemaUtils;
+import com.linkedin.venice.utils.Pair;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.function.BiFunction;
+import org.apache.avro.Schema;
+import org.apache.avro.generic.GenericRecord;
+
+
+/**
+ * Provides a semantic validation function to check whether an object adheres to a given schema.
+ * An object is considered valid if:
+ * <ol>
+ *   <li>It is null.</li>
+ *   <li>It is a nested type and equals its schema's default value.</li>
+ *   <li>It is a non-nested field and matches either the schema's default value or its type's default.</li>
+ * </ol>
+ *
+ * If the object is not using new semantic, we return False. Otherwise, we return True along with the message
+ * in errorMessage. The traverser should stop traversing the object if the return value is True.
+ */
+public class NewSemanticUsageValidator {
+  private static final ThreadLocal<String> errorMessage = ThreadLocal.withInitial(() -> "");
+
+  /**
+   * This is the validator function that we will pass into SchemaDiffTraverser to validate the two nodes
+   * with different schemas.
+   */
+  private final BiFunction<Object, Pair<Schema.Field, Schema.Field>, Boolean> SEMANTIC_VALIDATOR =
+      (object, schemasPair) -> {
+        Schema.Field currentField = schemasPair.getFirst();
+        Schema.Field targetField = schemasPair.getSecond();
+
+        // If object is null, return false, no need to validate more
+        if (object == null) {
+          return false;
+        }
+
+        // Default behavior, check if the value is non-default
+        return isNonDefaultValue(object, currentField, targetField);
+      };
+
+  /**
+   * Get the semantic validator function.
+   */
+  public BiFunction<Object, Pair<Schema.Field, Schema.Field>, Boolean> getSemanticValidator() {
+    return SEMANTIC_VALIDATOR;
+  }
+
+  /**
+   * General method to check if the value is non-default for the given field.
+   * @param object the value to check
+   * @param currentField the field to check
+   * @param targetField the target field to check
+   * @return true if the value is non-default, false otherwise
+   */
+  public boolean isNonDefaultValue(Object object, Schema.Field currentField, Schema.Field targetField) {
+    switch (currentField.schema().getType()) {
+      case UNION:
+        return isNonDefaultValueUnion(object, currentField);
+      case ENUM:
+        return isNewEnumValue(object, currentField, targetField);
+      case FIXED:
+        if (currentField.schema().getFixedSize() != targetField.schema().getFixedSize()) {
+          return returnTrueAndLogError(
+              String.format("Field %s: has different fixed size", formatFieldName(currentField.name())));
+        }
+        return false;
```

Comment:
Fixed is a tricky one; I believe with the current implementation, if the new field is a FIXED schema, we are enforcing the usage of new semantics.
Currently we are using the latest schema to serialize, and then do another round of deserialization and serialization.
If the codes put null to the new FIXED field, the first serialization will fail with NPE, so developers must put a fixed size array to this field.